lib/parsers/heif_parser.rb

```diff
@@ -0,0 +1,430 @@
+      item_info_end_pos = @buf.pos + item_info_entry_length - HEADER_LENGTH
+      version = read_int_8
+      safe_skip(@buf, 3) # 0 flags
+      case version
+      when 2
+        item_id = read_int_16
+      when 3
+        item_id = read_int_32
+      else
+        return # wrong version according to standards, hence return
+      end
+      safe_skip(@buf, 2) # not interested in the item_protection_index
+      item_type = read_string(4)
+      content_encoding = ''
+      if item_type == MIME_MARKER
+        content_encoding = read_string(item_info_end_pos - @buf.pos).delete!("\0") # remove the null-termination part for output visualization reason
+      end
+      @sub_items << {'item_id': item_id, 'item_type': item_type, 'content_encoding': content_encoding}
+      @buf.seek(item_info_end_pos) # we are not interested in anything else, go directly to the end of this 'infe' box
+    }
+  end
+
+  def read_nil_version_and_flag
+    safe_skip(@buf, 1) # version, always 0 in this current box
+    safe_skip(@buf, 3) # flags, always 0 in this current box
+  end
+
+  def read_primary_item_box
+    version = read_int_8
+    safe_read(@buf, 3) # flags, always 0 in this current box
+    @primary_item_id = if version == 0
+      read_int_16
+    else
+      read_int_32
+    end
+  end
+
+  # the ITEM_PROPERTIES_CONTAINER_BOX contains an implicitely 1-based index list of item properties.
+  # While parsing such box we are storing the properties with its own index.
+  # Reason behind is that the primary_item will be associated to some of these properties through the same index
+  # and in order to output relevant data from the format_parser we need all the properties associated to the primary_item.
+  # Hence the need of the association between an item and its properties, found in the ITEM_PROPERTIES_ASSOCIATION_BOX
+  def read_item_properties_box
+    ipco_length = read_int_32
+    return unless read_string(4) == ITEM_PROPERTIES_CONTAINER_BOX
+    read_item_properties_container_box(ipco_length)
+    read_int_32 # ipma_length
+    return unless read_string(4) == ITEM_PROPERTIES_ASSOCIATION_BOX
+    read_item_properties_association_box
+  end
+
+  def read_item_properties_container_box(box_length)
+    end_of_ipco_box = @buf.pos + box_length - HEADER_LENGTH
+    item_prop_length = read_int_32
+    item_prop_name = read_string(4)
+    item_prop_start_pos = @buf.pos
+    item_prop_index = 1
+    while @buf.pos < end_of_ipco_box
+      case item_prop_name
+      when IMAGE_SPATIAL_EXTENTS_BOX
+        read_nil_version_and_flag
+        width = read_int_32
+        height = read_int_32
+        @item_props[item_prop_index] = {
+          'type': IMAGE_SPATIAL_EXTENTS_BOX,
+          'width': width,
+          'height': height
+        }
+      when PIXEL_ASPECT_RATIO_BOX
+        h_spacing = read_int_32
+        v_spacing = read_int_32
+        pixel_aspect_ratio = h_spacing.to_s + '/' + v_spacing.to_s
```

Comment:
You could use interpolation to make this less verbose:

```suggestion
        pixel_aspect_ratio = "#{h_spacing}/#{v_spacing}"
```